Dockerfile

```diff
@@ -0,0 +1,298 @@
+RUN echo "Building airflow-build-image stage" \
+    echo "Base image: ${PYTHON_BASE_IMAGE}"; \
+    echo "Airflow version: ${AIRFLOW_VERSION}"; \
+    echo "Airflow git reference: ${AIRFLOW_GIT_REFERENCE}"; \
+    echo "Airflow org: ${AIRFLOW_ORG}"; \
+    echo "Airflow repo: ${AIRFLOW_REPO}"; \
+    echo "Airflow repo url: ${AIRFLOW_REPO_URL}"; \
+    echo "Airflow extras: ${AIRFLOW_EXTRAS}" ;\
+    echo "PIP version: ${PIP_VERSION}" ;\
+    echo "Cassandra concurrency: ${CASS_DRIVER_BUILD_CONCURRENCY}" ;\
+    echo
+
+# Make sure noninteractive debian install is used and language variables set
+ENV DEBIAN_FRONTEND=noninteractive LANGUAGE=C.UTF-8 LANG=C.UTF-8 LC_ALL=C.UTF-8 \
+    LC_CTYPE=C.UTF-8 LC_MESSAGES=C.UTF-8
+
+# Note missing man directories on debian-buster
+# https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=863199
+# Install basic apt dependencies
+RUN mkdir -pv /usr/share/man/man1 \
+    && mkdir -pv /usr/share/man/man7 \
+    && apt-get update \
+    && apt-get install -y --no-install-recommends \
+           apt-transport-https \
+           apt-utils \
+           build-essential \
+           ca-certificates \
+           curl \
+           gnupg \
+           dirmngr \
+           freetds-bin \
+           freetds-dev \
+           gosu \
+           krb5-user \
+           ldap-utils \
+           libffi-dev \
+           libkrb5-dev \
+           libpq-dev \
+           libsasl2-2 \
+           libsasl2-dev \
+           libsasl2-modules \
+           libssl-dev \
+           locales  \
+           lsb-release \
+           openssh-client \
+           postgresql-client \
+           python-selinux \
+           sasl2-bin \
+           software-properties-common \
+           sqlite3 \
+           sudo \
+           unixodbc \
+           unixodbc-dev \
+    && apt-get autoremove -yqq --purge \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+# Install MySQL client from Oracle repositories (Debian installs mariadb)
+RUN KEY="A4A9406876FCBD3C456770C88C718D3B5072E1F5" \
+    && GNUPGHOME="$(mktemp -d)" \
+    && export GNUPGHOME \
+    && for KEYSERVER in $(shuf -e \
+            ha.pool.sks-keyservers.net \
+            hkp://p80.pool.sks-keyservers.net:80 \
+            keyserver.ubuntu.com \
+            hkp://keyserver.ubuntu.com:80 \
+            pgp.mit.edu) ; do \
+          gpg --keyserver "${KEYSERVER}" --recv-keys "${KEY}" && break || true ; \
+       done \
+    && gpg --export "${KEY}" | apt-key add - \
+    && gpgconf --kill all \
+    rm -rf "${GNUPGHOME}"; \
+    apt-key list > /dev/null \
+    && echo "deb http://repo.mysql.com/apt/debian/ stretch mysql-5.7" | tee -a /etc/apt/sources.list.d/mysql.list \
+    && apt-get update \
+    && apt-get install --no-install-recommends -y \
+        libmysqlclient-dev \
+        mysql-client \
+    && apt-get autoremove -yqq --purge \
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
+
+# disable bytecode generation
+ENV PYTHONDONTWRITEBYTECODE=1
```

Comment:
Should we remove this variable after installation?

Reply:
Again - this is build image not the main image (see below comments separating those two images). This variable is only set for the build image, not for the main one. Thanks to that the .local folder is much smaller to copy - and the cache will be generated on the first run of python interpreter for the main image.

Reply:
> and the cache will be generated on the first run of python interpreter for the main image

I'd probably suggest that we have it written here -- that way the production image doesn't have to make any changes (it's often not possible, but it's nice to be able to run docker containers with a ro root disk).

I.e. it's nice to be able to run `docker diff airflow` and not see a bunch of .pyc files scattered about.

Reply:
I see later in the file you are explicitly deleting them -- did this env var not actually do anything anyway?

Reply:
It's not _quite_ the same, but this is almost like we are copying the .java files over and purposefully leaving the `.class` files out, to then recompile them later.

(Yes, it's just a cache in Python so this isn't quite the same I know).

Still, it seems like we are putting extra work in to every container that ever starts up :)

Reply:
This is really a trade-off between the size of the image and download bandwidth/time.

By deleting the .pyc files we get around 100 MB less in the layer. Those are binary files, so they do not compress well. I can get the exact numbers to see but I thought the size difference is quite significant.

I think it's milliseconds of .pyc generation (and it's also distributed over time) vs. downloading 100MB more or so. 

I can get the numbers in a moment so that we can make a better decision on that - both timing of pyc/no-pyc generation for typical airflow startup vs. size of the image.

Reply:
Without .pyc:

Airflow Layer = 419 MB, Compressed size = 99.44 MB
Total Image size = 837 MB. Compressed size = 237.2 MB

With .pyc:

Airflow Layer = 516 MB, Compressed size = 125 MB
Total image = 933 MB, Compressed size = 263.17 MB

So we save 26 MB of transfer and almost 100 MB of the image layer after decompression - 10% more or less of both compressed and uncompressed size. 

I think this is quite significant saving. 

I also measured startup time with and without .pyc and for webserver it was around a second (hard to say exactly).

and yes the PYTHONDONTWRITEBYTECODE did not work for install from sources. not sure yet why.

Reply:
WDYT @ashb @mik-laj @kaxil @turbaszek @feluelle -> should we leave or remove .pyc files ? It's a difficult decision to make :)

Reply:
I think let's leave it as it is for now. We can decide on it and update it later to unblock 1.10.10RC

Reply:
Okay yeah, that's big enough a difference to say leave it out.

For 3.8+ it might be worth looking at https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPYCACHEPREFIX to make it store the PYC files elsewhere in the tree (i.e. in to a `/tmp` somewhere, both at build image and prod image runtime)